src/EventListener/DeserializeListener.php

```diff
@@ -64,32 +66,43 @@ public function onKernelRequest(GetResponseEvent $event)
         $request->attributes->set(
             'data',
             $this->serializer->deserialize(
-                $request->getContent(), $attributes['resource_class'], $this->getFormat($request), $context
+                $request->getContent(), $attributes['resource_class'], $format, $context
             )
         );
     }
 
     /**
-     * Extracts the format form the Content-Type header and fallback to the format of the response.
+     * Extracts the format from the Content-Type header and check that it is supported.
      *
      * @param Request $request
      *
+     * @throws NotAcceptableHttpException
+     *
      * @return string
      */
     private function getFormat(Request $request) : string
     {
-        $format = $request->getContentType();
-        if (null !== $format && isset($this->formats[$format])) {
-            return $format;
+        $contentType = $request->headers->get('CONTENT_TYPE');
+        if (null === $contentType) {
+            throw new NotAcceptableHttpException('The "Content-Type" header must exist.');
         }
 
-        $requestFormat = $request->getRequestFormat();
-        if (isset($this->formats[$requestFormat])) {
-            return $requestFormat;
-        }
+        $format = $request->getFormat($contentType);
+        if (!isset($this->formats[$format])) {
+            $supportedMimeTypes = [];
+            foreach ($this->formats as $mimeTypes) {
+                foreach ($mimeTypes as $mimeType) {
```

Comment:
Anyway, I don't think the logic here is correct:

1. If the MIME type is null or unknown
2. If the Accept request header is not set
3. If there's a request format specified\*

\* Correct me if I'm wrong, but looking at
https://github.com/symfony/http-foundation/blob/472367d63660b5f1426b33b87ba0558b3487945a/Request.php#L1377-L1397
and referring to
http://symfony.com/doc/current/quick_tour/the_controller.html#using-formats
this doesn't seem like what we intend to check - the format **requirement**.